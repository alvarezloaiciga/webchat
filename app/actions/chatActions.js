import {reduxActionTypes as actionTypes} from 'Common/Constants';
import type {
  ChatInitializedStateType,
  Message,
  ChatMetadata,
  AttachmentError,
  QuiqObject,
} from 'types';
import type {PersistentData, Author} from 'quiq-chat/src/types';

export const setChatContainerHidden = (chatContainerHidden: boolean) => ({
  type: 'CHAT_CONTAINER_HIDDEN',
  chatContainerHidden,
});

export const setChatRegistrationField = (fieldId: string, fieldValue: any) => ({
  type: 'CHAT_REGISTRATION_FIELD_SET',
  fieldId,
  fieldValue,
});

export const setChatLauncherHidden = (chatLauncherHidden: boolean) => ({
  type: 'CHAT_LAUNCHER_HIDDEN',
  chatLauncherHidden,
});

export const setAgentsAvailable = (agentsAvailable: boolean) => ({
  type: 'AGENTS_AVAILABLE',
  agentsAvailable,
});

export const setIsAgentAssigned = (isAgentAssigned: boolean) => ({
  type: 'AGENT_ASSIGNED',
  isAgentAssigned,
});

export const setChatInitialized = (initializedState: ChatInitializedStateType) => ({
  type: 'CHAT_INITIALIZED_STATE',
  initializedState,
});

export const setTypingAuthorData = (author: Author | null) => ({
  type: 'UPDATE_TYPING_AUTHOR',
  author,
});

export const setMessageFieldFocused = (messageFieldFocused: boolean) => ({
  type: 'MESSAGE_FIELD_FOCUSED',
  messageFieldFocused,
});

export const updateChatConfigurationFromMetadata = (metadata: ChatMetadata) => ({
  type: 'UPDATE_CHAT_CONFIGURATION',
  configuration: {
    enableChatEmailTranscript:
      metadata.configs.CHAT_EMAIL_TRANSCRIPT && metadata.configs.CHAT_EMAIL_TRANSCRIPT.enabled,
    enableChatFileAttachments:
      metadata.configs.CHAT_FILE_ATTACHMENTS && metadata.configs.CHAT_FILE_ATTACHMENTS.enabled,
    enableManualConvoStart: metadata.configs.CHAT_MANUAL_START_CONVO_MODE,
    enableMobileChat: metadata.configs.CHAT_ALLOW_MOBILE,
    supportedAttachmentTypes:
      metadata.configs.ALLOWED_CONTENT_TYPES && metadata.configs.ALLOWED_CONTENT_TYPES.contentTypes,
    enableEmojis: metadata.configs.ENABLE_EMOJIS && metadata.configs.ENABLE_EMOJIS.enabled,
    playSoundOnNewMessage:
      metadata.configs.PLAY_SOUND_ON_NEW_MESSAGE &&
      metadata.configs.PLAY_SOUND_ON_NEW_MESSAGE.enabled,
    flashNotificationOnNewMessage:
      metadata.configs.FLASH_NOTIFICATION_ON_NEW_MESSAGE &&
      metadata.configs.FLASH_NOTIFICATION_ON_NEW_MESSAGE.enabled,
    registrationForm: metadata.registrationForm,
    registrationFormVersionId: metadata.registrationFormVersionId,
    menuOptions: metadata.configs.CHAT_MENU_OPTIONS,
    whitelistedDomains:
      metadata.configs.CHAT_WHITELISTED_DOMAINS &&
      metadata.configs.CHAT_WHITELISTED_DOMAINS.domains,
  },
});

export const updateChatConfigurationFromQuiqOptions = (quiqOptions: QuiqObject) => ({
  type: 'UPDATE_CHAT_CONFIGURATION',
  configuration: quiqOptions,
});

export const updatePersistentData = (persistentData: PersistentData) => ({
  type: 'CHAT_UPDATE_PERSISTENT_DATA',
  persistentData,
});

export const updatePlatformEvents = (platformEvents: Array<Event>) => ({
  type: 'UPDATE_PLATFORM_EVENTS',
  platformEvents,
});

export const updateTranscript = (transcript: Array<Message>) => ({
  type: 'UPDATE_TRANSCRIPT',
  transcript,
});

export const removeMessage = (messageId: string) => ({
  type: actionTypes.removeMessage,
  messageId,
});

export const setWelcomeFormRegistered = () => ({
  type: 'WELCOME_FORM_REGISTERED',
});

export const newWebchatSession = () => ({
  type: 'NEW_WEBCHAT_SESSION',
});

export const setUploadProgress = (messageId: string, progress: number) => ({
  type: actionTypes.uploadProgress,
  messageId,
  progress,
});

export const addAttachmentError = (attachmentError: AttachmentError) => ({
  type: actionTypes.addAttachmentError,
  attachmentError,
});

export const markSendFailure = (messageId: string, reason: string) => ({
  type: actionTypes.markSendFailure,
  messageId,
  reason,
});

export const addPendingAttachmentMessage = (
  tempId: string,
  contentType: string,
  blobUrl: string,
  fromCustomer: boolean,
) => {
  return {
    type: actionTypes.addPendingMessage,
    message: {
      id: tempId,
      localBlobUrl: blobUrl,
      localKey: tempId,
      type: 'Attachment',
      timestamp: Date.now(),
      authorType: fromCustomer ? 'Customer' : 'User',
      status: 'pending',
      contentType,
    },
  };
};

export const updatePendingAttachmentId = (tempId: string, newId: string) => ({
  type: actionTypes.updatePendingMessageId,
  tempId,
  newId,
});

export const setInputtingEmail = (inputtingEmail: boolean) => ({
  type: 'SET_INPUTTING_EMAIL',
  inputtingEmail,
});

export const setWindowScrollLockEnabled = (enabled: boolean) => ({
  type: 'SET_WINDOW_SCROLL_LOCK_ENABLED',
  enabled,
});

export const reloadApp = () => {
  // Reload the react app
};
