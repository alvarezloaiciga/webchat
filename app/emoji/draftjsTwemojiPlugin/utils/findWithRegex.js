const findWithRegex = (regex, contentBlock, callback) => {
  // Get the text from the contentBlock
  const text = contentBlock.getText();
  let matchArr;
  let start; // eslint-disable-line
  // Go through all matches in the text and return the indices to the callback
  // eslint-disable-next-line no-cond-assign
  while ((matchArr = regex.exec(text)) !== null) {
    start = matchArr.index;
    callback(start, start + matchArr[0].length);
  }
};

export default findWithRegex;
