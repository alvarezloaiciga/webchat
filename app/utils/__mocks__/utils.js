export const getBrowserName = () => 'Chrome';
export const getMajor = () => 60;
export const getDeviceType = () => 'PC';
export const getOSName = () => 'Windows';
export const getUAInfo = () => ({
  browser: {
    name: 'Chrome',
    major: 60,
  },
  os: {
    name: 'Mac OS',
    version: 10,
  },
  device: {
    model: 'iPad',
    vendor: 'Apple',
    type: 'tablet',
  },
  cpu: {
    architecture: 'amd64',
  },
  engine: {
    name: 'WebKit',
    version: 537.36,
  },
  navigator:
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.81 Safari/537.36',
});
export const compatibilityMode = () => false;
export const isIE9 = () => false;
export const isIE10 = () => false;
export const nonCompatibleBrowser = () => false;
export const supportsFlexbox = () => true;
export const supportsSVG = () => true;
export const displayError = (error: string) => error;
export const isIEorSafari = () => false;
