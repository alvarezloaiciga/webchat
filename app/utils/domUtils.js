import ReactDOM, {render} from 'react-dom';
import React from 'react';
import Routes from 'Routes';
import {quiqContainerId} from 'Common/Constants';
import {AppContainer} from 'react-hot-loader';
import {IntlProvider} from 'react-intl';
import Redbox from 'redbox-react';
import {Provider} from 'react-redux';

export const appIsMounted = (): boolean => {
  const container = document.getElementById(quiqContainerId);
  if (!container) return false;
  return container.childElementCount > 0;
};

export const destructApp = () => {
  ReactDOM.unmountComponentAtNode(document.getElementById(quiqContainerId));
};

export const constructApp = store => {
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <AppContainer errorReporter={Redbox}>
          <Routes />
        </AppContainer>
      </IntlProvider>
    </Provider>,
    document.getElementById(quiqContainerId),
  );

  if (module.hot) {
    module.hot.accept('Routes', () => {
      const NextApp = require('Routes').default; // eslint-disable-line global-require

      render(
        <Provider store={store}>
          <IntlProvider locale="en">
            <AppContainer errorReporter={Redbox}>
              <NextApp />
            </AppContainer>
          </IntlProvider>
        </Provider>,
        document.getElementById(quiqContainerId),
      );
    });
  }
};
