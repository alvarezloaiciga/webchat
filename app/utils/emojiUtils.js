import {emojiIndex} from 'emoji-mart';
import emojiRegexFactory from 'emoji-regex';
import invertBy from 'lodash/invertBy';
import type {EmojiMetadata, Emoji} from 'Common/Types';
import quiqOptions from 'Common/QuiqOptions';

const emojiRegex = emojiRegexFactory();

// Build lookup table of {unifiedCode : emojiId}
const emojiIdByUnified = {};
Object.keys(emojiIndex.emojis).forEach(id => {
  const unified = emojiIndex.emojis[id].unified.toLowerCase();
  emojiIdByUnified[unified] = id;
});

export const isSingleEmoji = (text: string): boolean => {
  // We know a string contains only one emoji iff
  // 1) An emoji regex finds only one match in the string, and
  // 2) The number of characters in that emoji === the number of characters in the string
  const emojis = text.match(emojiRegex);

  if (!emojis)
    return false;

  const emojiCharCount = emojis.reduce((count, e) => count + e.length, 0);
  return emojis.length === 1 && text.length === emojiCharCount;
};

export const convertUnicodeToEmojiObject = (u: string): ?Emoji => {
  const emojiId = Object.keys(emojiIndex.emojis).find(k => emojiIndex.emojis[k].native === u);
  return emojiIndex.emojis[emojiId];
};

// This function is used for filtering emojis in our picker, as well as redacting emojis prior to sending a message.
export const emojiFilter = (e: EmojiMetadata | Emoji | string) => {
  let emojiId;
  // emoji-mart will call this function when user is searching, but it only passes unicode string:
  if (typeof e === 'string') {
    emojiId =  emojiIdByUnified[e.toLowerCase()];
  } else {
    // NOTE: In emoji-mart land, id === short_codes[0]
    emojiId = e.id || e.short_names[0];
  }

  // Don't include any emojis that don't have id or shortcode (we don't know what they are)
  if (!emojiId)
    return false;

  const {includeEmojis, excludeEmojis} = quiqOptions;
  // Prioritize include over exclude. Only one of the should be defined by customer, however.
  if (Array.isArray(includeEmojis)) {
    return includeEmojis.includes(emojiId);
  }
  else if (Array.isArray(excludeEmojis)) {
    return !excludeEmojis.includes(emojiId);
  }

  // If neither exclude nor include was defined, always include emoji.
  return true;
};
