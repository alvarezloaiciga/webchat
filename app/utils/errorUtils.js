// @flow

import * as Utils from 'Common/Utils';

export const getMetadataForSentry = () => {
  return {
    inStandaloneMode: Utils.inStandaloneMode(),
  };
};
