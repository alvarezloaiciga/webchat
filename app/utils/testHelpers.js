// @flow
import type {Message, IntlObject, IntlMessage} from 'types';

export const getMockMessage = (
  // eslint-disable-line import/prefer-default-export
  id?: number = 0,
  overrides?: Message | {} = {},
) => {
  const message = {
    authorType: id % 2 === 0 ? 'Customer' : 'Agent',
    text: `text--${id}`,
    id: `id--${id}`,
    timestamp: id,
    type: 'Text',
  };

  return Object.assign(message, overrides);
};

/**
 * Stub out formatMessage so that it will return the defaultMessage with basic variable replacement
 */
const formatTestMessage = (msg: IntlMessage, values?: {[key: string]: string}) => {
  if (!values) {
    return msg.defaultMessage;
  }

  let returnValue = msg.defaultMessage;
  Object.keys(values).forEach(key => {
    if (values) {
      const value = values[key];
      returnValue = returnValue.replace(`{${key}}`, value);
    }
  });

  return returnValue;
};

export const TestIntlObject: IntlObject = {
  formatMessage: formatTestMessage,
  formatRelative: value => value.toString(),
  formatTime: timestamp => timestamp.toString(),
  formatDate: timestamp => timestamp.toString(),
};
