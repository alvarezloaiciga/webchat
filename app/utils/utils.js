// @flow
declare var Modernizr: Object;
import {UAParser} from 'ua-parser-js';
import type {BrowserNames, DeviceTypes, OSNames} from 'types';

const parser = new UAParser();

export const getBrowserName = (): BrowserNames => parser.getResult().browser.name;

export const getMajor = (): number => parseInt(parser.getResult().browser.major, 10);

export const getDeviceType = (): DeviceTypes => parser.getResult().device.type;

export const getOSName = (): OSNames => parser.getResult().os.name;

export const getUAInfo = () => parser.getResult();

export const compatibilityMode = () => {
  const compatList = [
    {
      name: 'IE',
      major: 10,
    },
  ];
  const {name, major} = parser.getResult().browser;
  return !!compatList.find(i => i.name === name && parseInt(major, 10) <= i.major);
};

export const isIE9 = () => getBrowserName() === 'IE' && getMajor() <= 9;
export const isIE10 = () => getBrowserName() === 'IE' && getMajor() === 10;

export const nonCompatibleBrowser = () => getBrowserName() === 'IE' && getMajor() < 9;
// Modernizr says IE10 doesn't support flexbox.
// It kind of does, at least for what we need it for... so go ahead and ignore Modernizr in that case
export const supportsFlexbox = () => isIE10() || Modernizr.flexbox;
