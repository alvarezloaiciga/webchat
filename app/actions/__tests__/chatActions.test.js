// @flow
jest.mock('utils/utils');
import * as chatActions from '../chatActions';
import {getMockMessage} from 'utils/testHelpers';
import {isIEorSafari} from 'utils/utils';

describe('chatActions', () => {
  describe('setChatHidden', () => {
    it('builds an action', () => {
      expect(chatActions.setChatHidden(true)).toMatchSnapshot();
    });

    describe('when in IE or Safari, not in standalone, and not hidden', () => {
      it('does not build an action', () => {
        (isIEorSafari: any).mockReturnValue(true);
        expect(chatActions.setChatHidden(false)).toBe(undefined);
      });
    });
  });

  describe('setChatInitialized', () => {
    it('builds an action', () => {
      expect(chatActions.setChatInitialized('initialized')).toMatchSnapshot();
    });
  });

  describe('setChatPopped', () => {
    it('builds an action', () => {
      expect(chatActions.setChatPopped(true)).toMatchSnapshot();
    });
  });

  describe('setAgentTyping', () => {
    it('builds an action', () => {
      expect(chatActions.setAgentTyping(true)).toMatchSnapshot();
    });
  });

  describe('updateTranscript', () => {
    it('builds an action', () => {
      expect(chatActions.updateTranscript([getMockMessage()])).toMatchSnapshot();
    });
  });

  describe('setWelcomeFormSubmitted', () => {
    it('builds an action', () => {
      expect(chatActions.setWelcomeFormSubmitted(true)).toMatchSnapshot();
    });
  });
});
