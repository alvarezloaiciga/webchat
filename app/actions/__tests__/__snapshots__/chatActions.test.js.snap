// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chatActions newWebchatSession builds an action 1`] = `
Object {
  "type": "NEW_WEBCHAT_SESSION",
}
`;

exports[`chatActions setAgentTyping builds an action 1`] = `
Object {
  "agentTyping": true,
  "type": "AGENT_TYPING",
}
`;

exports[`chatActions setChatContainerHidden builds an action 1`] = `
Object {
  "chatContainerHidden": true,
  "type": "CHAT_CONTAINER_HIDDEN",
}
`;

exports[`chatActions setChatInitialized builds an action 1`] = `
Object {
  "initializedState": "initialized",
  "type": "CHAT_INITIALIZED_STATE",
}
`;

exports[`chatActions setChatLauncherHidden builds an action 1`] = `
Object {
  "chatLauncherHidden": true,
  "type": "CHAT_LAUNCHER_HIDDEN",
}
`;

exports[`chatActions setWelcomeFormRegistered builds an action 1`] = `
Object {
  "type": "WELCOME_FORM_REGISTERED",
}
`;

exports[`chatActions updateTranscript builds an action 1`] = `
Object {
  "transcript": Array [
    Object {
      "authorType": "Customer",
      "id": "id--0",
      "text": "text--0",
      "timestamp": 0,
      "type": "Text",
    },
  ],
  "type": "UPDATE_TRANSCRIPT",
}
`;
